### ISSUE TITLE
Faulty istiod pod not pushing configs to sidecars

### ISSUE THREAD

**Issue by bharath-12345** [istio/istio#30539]
(**NOTE**: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>)

**Bug description**
One fine day, we were receiving a lot of alerts of 503s service unavailable for different services. Upon checking the services, we found that they themselves were in a healthy state. We suspected that istiod was not pushing configs to the sidecars. We were seeing the following error messages in certain new deployments during this time.
`2021-01-28T12:07:53.860590Z	warn	Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected`
When we ran `istioctl proxy-status`, we found that the command was hanging for a long time. When we re-ran it a couple of times, we got the below error message
```
Error: error execing into istiod-77cccf64c6-j7grj/istio-system discovery container: command terminated with exit code 255
Error: Get "http://127.0.0.1:15014/debug/syncz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```
Note: `istioctl proxy-config` worked fine
We analyzed the logs of this pod and there seemed to be nothing of interest. The pod was less chirpier then the other istiod pods. It had a lot of 
`2021-01-28T12:53:20.676648Z\tinfo\tads\tADS: \"192.168.50.35:59324\" sidecar~192.168.50.35~<pod_name>.<namespace>~<namespace>.svc.cluster.local-33 terminated rpc error: code = Canceled desc = context canceled\n
`
We saw logs like the below too
`"2021-01-31T06:08:07.583064Z\twarn\tunable to get node \"ip-192-168-34-49.ap-south-1.compute.internal\" for pod \"prometheus-operator-prometheus-node-exporter-kxjdq\": node \"ip-192-168-34-49.ap-south-1.compute.internal\" not found\n"`

`"2021-01-28T12:54:54.932183Z\tinfo\tads\tClient closed connection sidecar~192.168.33.195~<namespace>-8644d44-szlnc.<namespace>~<namespace>.svc.cluster.local-34\n"`

`2021-01-28T16:48:10.056650Z\tinfo\trootcertrotator\tRoot cert is not about to expire, skipping root cert rotation.\n
`
Once we killed this pod, a new pod came up post which all the configs were pushed correctly and services were able to discover each other. The issue had been resolved. We suspect that this faulty pod was not pushing configs to the sidecars properly due to which services were reporting 503s.

We would like to understand why this happened? And how can we avoid such scenarios in the future?

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**
Istiod should push the configs to the sidecars normally on time so that other services don't throw 503s for a service due to an out of data config.

**Steps to reproduce the bug**
Not sure

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
istioctl version --remote
`client version: 1.7.3
control plane version: 1.7.3
data plane version: 1.7.3 (382 proxies)`

kubectl version --short
`Client Version: v1.18.8
Server Version: v1.18.9-eks-d1db3c`

**How was Istio installed?**
Using Istio-operator. We had an istio operator pod which reconciled istio using the istiooperator resource.

**Environment where the bug was observed (cloud vendor, OS, etc)**
AWS EKS 1.18 with istio 1.7.3

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.
There is a lot of sensitive information here which would need a considerable amount of time censoring. Will add the results of istioctl bug-report if it is needed.

**Comment by howardjohn** [comment]
Moving slack discussion here, this is probably a deadlock. We need a goroutine dump. I know there was a patch around this, but I cannot find it.. I suspect this was fixed at some point though

**Comment by howardjohn** [comment]
Fix is in #26260, in 1.8 but not 1.7

**Comment by bharath-12345** [comment]
Thanks for the update on this @howardjohn . I have a question, if there was deadlock shouldn't the istio pod have restarted? We saw 0 restarts for the pod?

**Comment by howardjohn** [comment]
deadlock would not trigger a restart of istiod. It just means the mutex is
stuck

**Comment by bharath-12345** [comment]
This happened to us again today. Lot of services reported 503 and istioctl proxy-status was hanging. We were not able to identify the pod causing the issue as istioctl proxy-status was taking too long to finish. Due to a lot of system failures occuring, we restarted the istiod deployment. Weirdly, the issue occured again after restarting. istioctl proxy-status was hanging. When we restarted again it worked. 
I understand that the fix is in 1.8 but is there any way we can identify this faster?
